syntax="proto2";
package rong.Paxos;


enum OpCode {
    kPaxosLease             = 1;  // Lease选主使用
    kPaxosBasic             = 2;  // 普通用
};

message Entry {
    required uint64 instance_id = 1;
    required uint64 node_id     = 2; //发起proposal的原始node_id，用于客户端检测自己的instance_id是否已经被占用
    required bytes  data        = 3;

    optional bytes  context     = 6;
};

message ApplyResponse {
    required uint64 instance_id = 1;
    required uint64 node_id     = 2;
    optional int32  code        = 3; // 执行返回码
    optional bytes  data        = 4;

    optional bytes  context     = 6;
};


// PaxosLease

enum LeaseMessageType {

    kLPrepareRequest        = 1;
    kLPrepareRejected       = 2;
    kLPreparePreviouslyAccepted = 3;
    kLPrepareCurrentlyOpen  = 4;

    kLProposeRequest        = 5;
    kLProposeRejected       = 6;
    kLProposeAccepted       = 7;
    
    kLProposeLearnOpen      = 8;
};

message LeaseMessage {

    required LeaseMessageType  type = 1;
    required uint64     node_id     = 2;
    required uint64     proposal_id = 3;
    required uint64     instance_id = 4;
    required uint64     log_last_index = 5;    // 本地最大的日志条目数目

    optional uint64     accepted_proposal_id = 6;
    optional uint64     lease_owner = 7;
    optional uint64     local_expire_time = 8;
};


// PaxosBasic
enum BasicMessageType {

    kBPrepareRequest        = 1;
    kBPrepareRejected       = 2;
    kBPreparePreviouslyAccepted = 3;
    kBPrepareCurrentlyOpen  = 4;    // 之前没有收到accpet，可以让Proposer指定

    kBProposeRequest        = 5;
    kBProposeRejected       = 6;
    kBProposeAccepted       = 7;

    kBProposeLearnValue     = 8;    // 当Proposer所在节点达成一致性后，对应的Learner会向所有节点发送Learn消息
    kBProposeLearnResponse  = 9;

    kBProposeChosenValue    = 10;   // 对于缺少的value，主动向日志更多的Node请求学习
    kBProposeChosenResponse = 11;   // 返回信息

    kBProposeStartCatchup   = 12;   // 快照

    // 当Prepare、Propose消息带有的instance_id过旧的时候就使用该消息返回chosen value
    kBMandatoryRejectChonsen = 20;  
};


message BasicMessage {

    required BasicMessageType  type = 1;
    required uint64     node_id     = 2;
    required uint64     proposal_id = 3;
    required uint64     instance_id = 4;       // 不同的消息类型有不同的含义
    required uint64     log_last_index = 5;    // 本地最大的日志条目数目
    
    optional uint64     accepted_proposal_id = 6;
    optional uint64     promised_proposal_id = 7;
    optional bytes      value       = 8;
    
    repeated bytes      chosen_values = 9;     // 当发起ChosenValue的时候，批量返回已经chosen的value
};

